Creating more granular configurations for IDE and CLI

joshbw opened the pull request:
Adding better configuration options for the IDE and CLI - specifically better ignore file functionality, the ability to only run certain rules (by ID or tag) or exclude certain rules (by ID or tag).  For the CLI, adding ability to control what objects go into various output formats

joshbw committed:
Made the file ignore functionality more robust

scovetta commented on package.json:
```diff
@@ -120,27 +120,43 @@
+            "*/logs/*",
+            "*/NuGet/*",
+            "*/.cache/*",
+            "*/tests/*",
```
Suggest either making these case insensitive, or even better, regular expressions.

joshbw replied:
regular expressions would actually make my life easier, since behind the scenes I convert them to RegExes for processing (initially did globs, but the glob libraries were so bloody slow and I didn't want to recreate them, so just did the wild cards you see).  Do you think requiring a person to be familiar with regular expressions to add to the list is the right usability choice?  Clearly for those of us working on DevSkim regex familiarity is a must, and it would provide a lot of power to people, but for novices (and we do hope novice devs are trying to find security flaws) do you think its ok to ask them to use regular expressions?

case insensitive would be super easy on my part (add an i flag) - do you think there are any scenarios in the *nix world where people would be annoyed by that?

scovetta replied:
You could go with a schema like `/.*\.cache.*/i` meaning "anything that contains .cache, case insensitive", so it's all in one. I think most developers would either be familiar or could learn what they mean pretty easily.

joshbw replied:
Works for me, and in the documentation for the settings I can link to scriptular.com/ to help newbies (and me) make/test their regex.  I'll try to make the change today or Monday (we have an adventure tomorrow).  Also, you and I suck at work/life balance

joshbw replied:
Hey, first pass, what do you think of these Regexes to cover the ignore scenarios:

/\.(exe|dll)$/i
/\.(png|jpg|jpeg|gif|psd|ico|mp3|mpeg)$/i
/\.(zip|tar|gz|rar)$/i
/(\/|\\)(out|bin)(\/|\\)/i
/(\/|\\)(test|test|\_\_tests__|\_\_mocks__)(\/|\\)/i
/(\/|\\)(\.git|git)(\/|\\)/i
/\.(git|git[^\.\\\/]*)$/i
/-lock[^\.\\\/]*|\.lock$/i
/(\/|\\)(\.vscode|\.cache|logs)(\/|\\)/i
/(\/|\\)(nuget|node_modules)(\/|\\)/i
/\.(log|sarif|test)$/i

scovetta replied:
Couple changes, additions

```
/\\.(exe|dll|so|dylib|bin|so\\..*)$/i
/\\.(png|jpg|jpeg|gif|psd|ico|mp3|mpeg|bmp)$/i
/\\.(zip|tar|gz|rar|jar|gz|7z|bz|bz2|gzip|cab|war|xz|nupkg|gem|egg)$/i
/\\.(sqlite3)$/i
/[/\\](out|bin)[/\\]/i
/[/\\](tests?|unittests?|__tests__|__mocks__)[/\\]/i
/[/\\](\\.git|git)[/\\]/i
/\\.(git|git[^.\/])$/i   <-- 
/-lock[^\\\./]|\\.lock$/i
/[/\\](\\.vscode|\\.cache|logs)[/\\]/i
/[/\\](nuget|node_modules)[/\\]/i
/\\.(log|sarif|test)$/i
/\\.(py[cod]$/i
/[/\\]__pycache__[/\\]/i
```

joshbw replied:
the bracket format is way better way of handling slashes.  should \_\_pycache__ be lumped in with node_modules/nuget

joshbw committed:
Refactoring for multipe output formats

joshbw committed:
Oops, overwrote the %SRCROOT% change

joshbw committed:
Added a max file size setting

joshbw committed:
clean up LGTM recommendations to keep debt down

joshbw committed:
resolve merge conflicts

joshbw commented:
@scovetta this branch now has a max file size setting, set to 100KB.  I've also change the command line behavior a bit, but your old command line will keep working (changed it to maintain compat).  The behavior for analyze it now looks like (also will apply to settings & rules, but they aren't as far along in implementation):

DevSkim analyze -o <filename>.<sarif | html | csv | txt>  //will infer the format of the output from the file extension

DevSkim analyze -f <s | t | h | c> //will use sarif, text, html, or csv and output to the console.  you can also spell out the formats instead of using the first letter, and case doesn't matter.

DevSkim analyze //outputs as text to the console

DevSkim analyze -f <s | t | h | c> -o //will output in the given format, to a devskim_results.<file extension>

DevSkim analyze -o //will output as text to the devskim_results.txt

joshbw committed:
object renaming for clarity, comments

joshbw committed:
more refactoring to minimize duplicate code

joshbw committed:
Change to ignores to use RegExes

Commit:
Use surrounding lines as snippet text.
Instead of using only the matched text, I updated the DevSkim worker
to include two lines before and after the line that the match starts
on, limited to 80 characters per line (for minimized code).

Commit:
Fix for missed files when target path starts with ./ or .\

joshbw committed:
Merge pull request #42 from microsoft/user/miscovet/snippets

Improve snippets; CLI target directory starting with ./

joshbw committed:
cleaned up regexes for ignores

joshbw committed:
Changes the name of IgnoreFiles setting

joshbw commented:
Merged PR, to get the existing changes into Master before I start really mucking things up
